### PR TITLE
Improve voice loading times

### PR DESCRIPTION
**Pull request by forslund**
This PR intends to improve the loading times of voice files (.flitevox). The new code is not active by default the option `--enable-voice-load-opt` needs to be added to configure for them to take effect. When this flag is enabled I see an improvement in loading speed of 20-25% (when running a series of 100 loads). I've added `time_voice_load` to the testsuite to check the load times (roughly).

My approach has been to try to reduce the number of calls to `fread()` and to limit the number of context switches into kernel mode.

The two most notable changes are
- reducing calls to `fread()` in `cst_read_tree_nodes()` (10% improvement)
- Increasing the vbuf for reads to 64k, this reduces context switches for reads by 90% (additional 5% improvement in time)

Many other tiny changes each reducing the load time by a couple of percent each contributes to the rest of the time improvements.

Currently I'm working on a block allocator to reduce context switches for memory allocation but I still have to fix some issues with that (make sure alignments are correct).

I haven't had the possibility to try this on a raspberry pi, the increased `vbuf` might make a bigger difference on such a system.

### CODE REVIEW

**Comment by forslund** [comment]
@el-tocino took the time to run this on a device similar to the raspberry pi and after some issues with swapping an improvement similar to mine could be observed.

However when the voice file (in this case *mycroft_voice_4.0.flitevox*) is completely uncached by the kernel the delay for the file operations are an order of magnitude larger than the improvements reducing the effectiveness of the optimization attempt to a couple of percent.

**Comment by forslund** [comment]
Now that Travis has run the os X build there is an actual issue. I'll see if I can fix it.

**Comment by forslund** [comment]
There, it's passing. I used the work around described in https://gist.github.com/jbenet/1087739

I haven't tested been able to test it properly since I have no computer running os X

**Comment by zeehio** [comment]
Sorry for not replying for a long time.

To me this looks good and can be merged. My only question is that I don't see why the optimization should be hidden behind a `--enable` option, since it seems that performance does not decrease in any case. Why don't we drop that option and always use your optimization by default?  In any case, this looks good to be merged to me. 

I don't know how far did you go with the block allocator and the alignment issues, but I bet it can provide much larger improvements. Another option if the alignment issues are annoying would be to have different memory blocks per data type to ensure alignment.

**Comment by forslund** [comment]
Yeah, the flag was mainly added to test the differences, I'll make an update and remove it.

I know I did a block-allocator at one point, and I have a separate branch for that (somewhere)...I think I wanted this merged separately for some reason but I can't quite remember.

File: src/cg/cst_cg_map.c [hunk 1]
```diff
@@ -77,16 +77,22 @@ cst_cg_db *cst_cg_load_db(cst_voice *vox, cst_file fd)
     cst_cg_db *db = cst_alloc(cst_cg_db, 1);
     int i;
     uint32_t elements[2];
-    uint32_t load_buff[4];
+    struct load_buff_s {
```
Comment (zeehio) [inline src/cg/cst_cg_map.c]: Relying on `sizeof(a structure)` is risky. We don't know were mimic may end up being used and what alignment issues we may face. Given that this happens just once (not any inner loop) what do you think about using two `cst_fread`, one for the integers and one for the floats?

Comment (forslund) [inline src/cg/cst_cg_map.c]: A good idea.

I was mainly testing my options with this psuh. Frankly I didn't think this would pass Travis.

**Comment by forslund** [comment]
Updated previous commit according to your suggestion.

**Comment by zeehio** [comment]
Merged! :-)

**Comment by forslund** [comment]
Then I'll move on to the block allocator :)

**Pull request merged**